### ISSUE TITLE
Backend readonly mode does not work for all widgets

### ISSUE THREAD

**Issue by netzarbeiter** [contao/contao#4868]
### Affected version(s)

4.9 & 4.13

### Description

```
contao:
    localconfig:
        timeZone: 'Europe/Berlin'
```

`timeZone` is not displayed as readonly field.

![bug](https://user-images.githubusercontent.com/1942550/175317027-e3548258-8082-48ab-84df-0c308504ed6d.png)

`defaultChmod` is not displayes ad readonly.

![bug](https://user-images.githubusercontent.com/1942550/175322767-da24e4b5-fece-4429-8b43-5bef0ba128ef.png)

**Comment by fritzmg** [comment]
This also applies to widgets in 4.9, not just 4.13.

**Comment by aschempp** [comment]
see https://github.com/contao/contao/pull/4870

**Comment by fritzmg** [comment]
The following widgets do not (fully) support `readonly => true`:

* `checkbox`
* `checkboxWizard`
* `chmod`
* `fileTree`
* `imageSize`
* `inputUnit` (the select is not disabled)
* `metaWizard` (removing still works, but the add button is disabled)
* `moduleWizard`
* `optionWizard`
* `pageTree`
* `picker`
* `radio`
* `sectionWizard`
* `select` with `chosen => true`
* `tableWizard` (you can still add/delete rows and columns, but you cannot edit cells)
* `timePeriod` (the select is not disabled)
* `trbl` (the select is not disabled)

**Comment by ausi** [comment]
> The following widgets do not (fully) support `readonly => true`:

I just tried to fix this but found out that there is no `readonly` for many form inputs, MDN writes:

> Only text controls can be made read-only, since for other controls (such as checkboxes and buttons) there is no useful distinction between being read-only and being disabled, so the readonly attribute does not apply.

We could implement the functionality ourselfs pretty easily with `$this->readonly ? ' onclick="return false"' : ''` but I’m not sure if this is needed.

**Comment by leofeyer** [comment]
If the input field does not support readonly, we do not have to, either, in my opinion.

**Comment by ausi** [comment]
Currently we half-support it for checkboxes, any change gets lost when saving, but the user can not see that directly.

I think we should either fully support it for all fields, or not support it for non-text fields (including the save mechanism).

**Comment by aschempp** [comment]
shouldn't we simply convert `readonly` to `disabled` for checkboxes?

**Comment by ausi** [comment]
> shouldn't we simply convert `readonly` to `disabled` for checkboxes?

IMO no, `readonly` has a different meaning (the input field can still be focused for example)

**Comment by aschempp** [comment]
> > shouldn't we simply convert `readonly` to `disabled` for checkboxes?
> 
> IMO no, `readonly` has a different meaning (the input field can still be focused for example)

yes, for input fields that **do** support `readonly`. But for the ones that do not (like a checkbox), converting it to `disabled` would solve the issue?

**Comment by ausi** [comment]
> But for the ones that do not (like a checkbox), converting it to `disabled` would solve the issue?

I don’t think so. `disabled` is just too different to me, as it would for example also not be sent by the browser anymore.

I think we should either fully support it for all fields, or not support it for non-text fields (including the save mechanism).

**Comment by aschempp** [comment]
How can we _fully support it for all fields_ if the HTML standard does not?

**Comment by ausi** [comment]
> How can we _fully support it for all fields_ if the HTML standard does not?

See https://github.com/contao/contao/issues/4868#issuecomment-1264333797 for checkboxes `onclick="return false"` should be sufficient.

**Comment by ausi** [comment]
> I think we should either fully support it for all fields, or not support it for non-text fields (including the save mechanism).

As discussed today we do want to support `readonly` for all field types, either by implementing a read-only mode ourselves (like `onclick="return false"`) or by mapping it to `disabled`.
